components/sup/src/http_gateway.rs

```diff
@@ -113,135 +112,193 @@ impl fmt::Display for ListenAddr {
     }
 }
 
-struct ManagerFs;
+#[derive(Default, Serialize)]
+struct HealthCheckBody {
+    status: String,
+    stdout: String,
+    stderr: String,
+}
 
-impl typemap::Key for ManagerFs {
-    type Value = manager::FsCfg;
+impl Into<StatusCode> for HealthCheck {
+    fn into(self) -> StatusCode {
+        match self {
+            HealthCheck::Ok | HealthCheck::Warning => StatusCode::OK,
+            HealthCheck::Critical => StatusCode::SERVICE_UNAVAILABLE,
+            HealthCheck::Unknown => StatusCode::INTERNAL_SERVER_ERROR,
+        }
+    }
 }
 
-pub struct Server(Iron<Chain>, ListenAddr);
+struct AppState {
+    fs_cfg: Arc<manager::FsCfg>,
+    gateway_state: Arc<RwLock<manager::GatewayState>>,
+}
 
-impl Server {
-    pub fn new(manager_state: Arc<manager::FsCfg>, listen_addr: ListenAddr) -> Self {
-        let mut r = Router::new();
-        if !feat::is_enabled(feat::RedactHTTP) {
-            r.get("/butterfly", butterfly, "butterfly");
-            r.get("/census", census, "census");
+impl AppState {
+    fn new(fs_cfg: Arc<manager::FsCfg>, gs: Arc<RwLock<manager::GatewayState>>) -> Self {
+        AppState {
+            fs_cfg: fs_cfg,
+            gateway_state: gs,
         }
+    }
+}
 
-        r.get("/", doc, "doc");
-        r.get("/services", services, "services");
-        r.get("/services/:svc/:group", service, "services_svc_group");
-        r.get(
-            "/services/:svc/:group/:org",
-            service,
-            "services_svc_group_org",
-        );
-        r.get(
-            "/services/:svc/:group/config",
-            config,
-            "services_svc_group_config",
-        );
-        r.get(
-            "/services/:svc/:group/health",
-            health,
-            "services_svc_group_health",
-        );
-        r.get(
-            "/services/:svc/:group/:org/config",
-            config,
-            "services_svc_group_org_config",
-        );
-        r.get(
-            "/services/:svc/:group/:org/health",
-            health,
-            "services_svc_group_org_health",
-        );
+pub struct Server;
 
-        let mut chain = Chain::new(r);
-        chain.link(persistent::Read::<ManagerFs>::both(manager_state));
-        Server(Iron::new(chain), listen_addr)
+impl Server {
+    pub fn run(
+        listen_addr: ListenAddr,
+        fs_cfg: Arc<manager::FsCfg>,
+        gateway_state: Arc<RwLock<manager::GatewayState>>,
+    ) {
+        thread::spawn(move || {
+            let sys = actix::System::new("sup-http-gateway");
+
+            server::new(move || {
+                let app_state = AppState::new(fs_cfg.clone(), gateway_state.clone());
+                App::with_state(app_state).configure(routes)
+            }).bind(listen_addr.to_string())
+            .unwrap()
```

Comment:
What's the desired behavior here?  Should the supervisor `panic!` if the address can't be bound to?  Should it do a graceful shutdown via `shutdown`, and if so, does this need its own `ShutdownReason`?

Reply:
I don't think the shutdown necessarily needs to be graceful, but I think it should be relatively immediate and in such a way that the launcher won't restart (having [`run`](https://github.com/habitat-sh/habitat/blob/master/components/sup/src/manager/mod.rs#L649) return an `Err` should suffice, but panicking won't). Since this shutdown doesn't need to be graceful, I don't know that you'll need to involve `ShutdownReason` at all, but the tricky part will be determining that the thread has panicked at all.

I couldn't find a good, simple approach in my small amount of googling. It may require sending a signal to the process in the even that `bind` fails.

Reply:
I think the cleanest approach may be to pass something like a `Mutex<Option<Result>>` into the context where `bind` is called so the outer context of `http_gateway::Server::run` can check it and supply a meaningful return value to _it's_ caller.

`None` would imply that the `bind` thread hadn't run yet and the `Result` can indicate whether the gateway started up successfully.

Reply:
@baumanj That's a cool idea 👍

Reply:
I decided to tackle this by attempting to connect to our HTTP listen address before spawning the background thread to bind. If the connection succeeds, it means something's already listening on that address, so I return an `Err` from `run()`, and everything halts with an error message. If the connection fails, it means we're good to bind, and everything continues. Check out my last commit and let me know what you guys think. It felt a little simpler than trying to coordinate between threads.